fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py

```diff
@@ -0,0 +1,767 @@
+                self.node.get_logger().info(
+                    f'[{self.name}] Stop requested from RMF!')
+                self.request_stop(mission)
+                self.mission = None
+
+    @parallel
+    def request_stop(self, mission: MissionHandle):
+        if mission is not None:
+            with mission.mutex:
+                if (mission.mission_queue_id is not None and
+                        not mission.do_not_cancel):
+                    self.api.mission_queue_id_delete(mission.mission_queue_id)
+
+    def localize(self, estimate, execution):
+        self.mission = MissionHandle(execution, localize=True)
+        self.mission.do_not_cancel = True
+        self.request_localize(estimate, self.mission)
+
+    @parallel
+    def request_localize(self, estimate, mission: MissionHandle):
+        retry_count = 10
+        count = 0
+        mission_queue_id = None
+        while count < retry_count and not mission_queue_id:
+            count += 1
+
+            # If this localize request is being made for a map switch, and a
+            # MiR position is provided in the fleet config for this RMF lift
+            # waypoint, we can localize to that position directly
+            mir_position = None
+            if estimate.inside_lift:
+                lift_name = estimate.inside_lift.name
+                this_lift = self.lift_positions.get(lift_name)
+                if this_lift and this_lift.get(estimate.map):
+                    mir_position = this_lift[estimate.map].get('name')
+
+            try:
+                mission_queue_id = self.api.localize(
+                    estimate.map, estimate.position, estimate.graph_index,
+                    mir_position
+                )
+                if mission_queue_id is not None:
+                    self.node.get_logger().info(
+                        f'[{self.name}] Localize mission is successfully '
+                        f'queued!')
+                    break
+            except Exception as err:
+                self.node.get_logger().info(
+                    f'Failed to localize on map {estimate.map}: {err}. '
+                    f'Retrying...'
+                )
+            time.sleep(1)
+
+        if not mission_queue_id:
+            self.node.get_logger().info(
+                f'[{self.name}] Failed to localize on map {estimate.map}. '
+                f'Maximum localize retries exceeded!'
+            )
+            mission.execution.finished()
+            mission_queue_id = None
+            return
+
+        mission.set_mission_queue_id(mission_queue_id)
+
+    def perform_action(self, category, description, execution):
+        if self.current_action:
+            # Should not reach here, but we log an error anyway
+            self.node.get_logger().info(
+                f'Robot [{self.name}] is busy with another perform action! '
+                f'Ignoring new action [{category}]')
+            execution.finished()
```

Comment:
Triggering `execution.finished()` here is pretty questionable. It means RMF will behave as though the newly requested action was completed even though it wasn't even performed.

It's true that this situation shouldn't generally happen, but I think if we need to consider it then we should preempt the ongoing action by simply pretending it isn't happening and performing the new action as though everything is normal. But maybe log an error message to indicate that something unusual has happened.

Reply:
Thanks for pointing this out. It actually helped me find an edge case where this could happen in the current code. In cart delivery, if the category supplied is invalid, [`execution.finished()` is called](https://github.com/open-rmf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_cart_delivery.py#L151), but the robot adapter's [`self.current_action` is created](https://github.com/open-rmf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py#L756) either way.
This should not happen as category matching was performed before creating the plugin, and additionally with your suggestion for `supports_action` I believe it is safe to remove this default case and the `execution.finished()` call.

I also wonder if it would be helpful to add a `self.abandon_action` flag in MirAction. If it is called, we will return `True` in `update_action()` and the robot adapter would set its `self.current_action` to `None`.

Lastly to clarify, do you mean that in this scenario we should let `self.current_action = None` and accept the new `perform_action` received?

Reply:
After working on the remaining refactor, I don't have very strong reasons to add an `end_action` flag in MirAction.

Regarding this unlikely scenario, I've modified the behavior to accept the new incoming actions, and end the current action in [9610e97](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/9610e9788b5ee3f9caca31d1b69b2b3ef53235a3)